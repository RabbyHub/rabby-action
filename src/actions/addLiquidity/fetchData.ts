import PQueue from 'p-queue';
import {
  AddLiquidityRequireData,
  FetchActionRequiredData,
  ParsedActionData,
} from '../../types';
import { waitQueueFinished } from '../../utils/waitQueueFinished';
import BigNumber from 'bignumber.js';

export const fetchDataAddLiquidity: FetchActionRequiredData<{
  receiver: string;
  exchange_rate: string;
  min_exchange_rate: string;
  max_exchange_rate: string;
  token0: any;
  token1: any;
}> = async (options, likeAction) => {
  const { actionData, apiProvider, sender, walletProvider, chainId } = options;

  const action =
    likeAction || (<ParsedActionData<'transaction'>>actionData).addLiquidity;

  if (!action || options.type === 'text') {
    return {};
  }

  const { receiver, exchange_rate, token0, token1 } = action;

  const fair_exchange_rate = new BigNumber(token0.price).div(token1.price);

  const diff = +new BigNumber(exchange_rate)
    .minus(fair_exchange_rate)
    .abs()
    .div(fair_exchange_rate)
    .times(100)
    .toFixed(4);

  const receiverInWallet = await walletProvider.hasAddress(receiver);

  const id =
    options.type === 'transaction'
      ? options.tx.to
      : options.actionData.contractId;
  if (!id) {
    return {};
  }

  const result: AddLiquidityRequireData = {
    id,
    protocol: null,
    bornAt: 0,
    rank: null,
    sender,
    receiverInWallet,
    hasInteraction: false,
    receiver,
    poolRate: exchange_rate,
    token0,
    token1,
    marketRate: fair_exchange_rate.toFixed(),
    diff,
  };
  const queue = new PQueue();

  queue.add(async () => {
    const contractInfo = await apiProvider.getContractInfo(id, chainId);
    if (!contractInfo) {
      result.rank = null;
    } else {
      result.rank = contractInfo.credit.rank_at;
      result.bornAt = contractInfo.create_at;

      result.protocol = contractInfo.protocol;
    }
  });

  queue.add(async () => {
    const hasInteraction = await apiProvider.hasInteraction(
      sender,
      chainId,
      id
    );
    result.hasInteraction = hasInteraction.has_interaction;
  });

  await waitQueueFinished(queue);

  return result;
};
