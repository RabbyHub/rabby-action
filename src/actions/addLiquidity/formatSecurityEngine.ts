import {
  AddLiquidityRequireData,
  FormatSecurityEngineContext,
} from '../../types';

export const formatSecurityEngineAddLiquidity: FormatSecurityEngineContext =
  async (options) => {
    const { actionData, requireData, chainId } = options;

    if (!actionData.addLiquidity || !chainId) {
      return {} as any;
    }

    const data = requireData as AddLiquidityRequireData;

    return {
      addLiquidity: {
        receiverMatch:
          data.receiver?.toLowerCase() === data.sender.toLowerCase(),
        diff: data.diff,
      },
    };
  };
