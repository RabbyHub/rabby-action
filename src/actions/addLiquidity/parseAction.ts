import { ParseAction } from '../../types';
import { TokenItem } from '@rabby-wallet/rabby-api/dist/types';

interface AddLiquidityActionData {
  receiver: string;
  exchange_rate: string;
  min_exchange_rate: string;
  max_exchange_rate: string;
  token0: TokenItem;
  token1: TokenItem;
}

export const parseActionAddLiquidity: ParseAction = (options) => {
  const { data } = options;

  if (data?.type !== 'add_liquidity') {
    return {};
  }

  const {
    receiver,
    exchange_rate,
    min_exchange_rate,
    max_exchange_rate,
    token0,
    token1,
  } = data.data as unknown as AddLiquidityActionData;

  return {
    addLiquidity: {
      receiver,
      exchange_rate,
      min_exchange_rate,
      max_exchange_rate,
      token0,
      token1,
    },
  };
};
