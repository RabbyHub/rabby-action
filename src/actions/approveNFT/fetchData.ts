import PQueue from 'p-queue';
import { ApproveNFTRequireData, FetchActionRequiredData } from '../../types';
import { waitQueueFinished } from '../../utils/waitQueueFinished';
import { fetchHasInteraction } from '../../fetches/fetchHasInteraction';

export const fetchDataApproveNFT: FetchActionRequiredData<{
  spender: string;
}> = async (options, likeAction) => {
  const queue = new PQueue();
  const { sender, apiProvider, chainId, actionData, extraActionDataState } =
    options;
  const action = likeAction || actionData.approveNFT;

  if (!action) {
    return {};
  }

  const result: ApproveNFTRequireData = {
    isEOA: false,
    contract: null,
    riskExposure: 0,
    rank: null,
    bornAt: 0,
    protocol: null,
    isDanger: false,
    tokenBalance: '0',
    hasInteraction: null,
    extraState: {
      hasInteraction: () => undefined,
    },
  };

  queue.add(async () => {
    const contractInfo = await apiProvider.getContractInfo(
      action.spender,
      chainId
    );
    if (!contractInfo) {
      result.isEOA = true;
      result.rank = null;
    } else {
      result.rank = contractInfo.credit.rank_at;
      result.riskExposure = contractInfo.top_nft_spend_usd_value;
      result.bornAt = contractInfo.create_at;
      result.isDanger =
        contractInfo.is_danger.auto || contractInfo.is_danger.edit;
      result.protocol = contractInfo.protocol;
    }

    if (result.isEOA) {
      queue.add(async () => {
        const { desc } = await apiProvider.addrDesc(action.spender);
        result.bornAt = desc.born_at;
      });
    }
  });

  fetchHasInteraction({
    apiProvider,
    sender,
    chainId,
    spender: action.spender,
    extraActionDataState,
    queue,
    result,
  });

  await waitQueueFinished(queue);
  return result;
};
