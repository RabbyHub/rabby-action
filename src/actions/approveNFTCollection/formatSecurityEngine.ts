import {
  ApproveNFTRequireData,
  FormatSecurityEngineContext,
} from '../../types';

export const formatSecurityEngineApproveNFTCollection: FormatSecurityEngineContext =
  async (options) => {
    const { actionData, requireData, chainId, provider } = options;

    if (!actionData.approveNFTCollection || !chainId) {
      return {};
    }
    const data = requireData as ApproveNFTRequireData;
    const { spender } = actionData.approveNFTCollection;
    return {
      collectionApprove: {
        chainId,
        spender,
        isEOA: data.isEOA,
        riskExposure: data.riskExposure,
        deployDays: provider.getTimeSpan(
          Math.floor(Date.now() / 1000) - data.bornAt
        ).d,
        isDanger: !!data.isDanger,
      },
    };
  };
