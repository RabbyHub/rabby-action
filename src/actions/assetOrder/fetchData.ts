import { FetchActionRequiredData } from '../../types';
import { fetchDataCommon } from '../common/fetchData';

export const fetchDataAssetOrder: FetchActionRequiredData = async (options) => {
  let action;

  if (options.type === 'transaction' && options.actionData.assetOrder) {
    action = {
      receiver: options.tx.to,
    };
  }
  if (
    options.type === 'typed_data' &&
    options.actionData.assetOrder &&
    options.actionData.contractId
  ) {
    action = {
      receiver: options.actionData.contractId,
    };
  }

  if (!action || !options.chainId) {
    return {};
  }

  return fetchDataCommon(options, action);
};
