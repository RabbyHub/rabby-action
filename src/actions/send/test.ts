import {
  apiProvider,
  ETH_CHAIN_ID,
  formatProvider,
  ORIGIN,
  SENDER,
  walletProvider,
} from '../../../__mocks__';
import { ParsedTransactionActionData } from '../../types';
import { fetchDataSend } from './fetchData';
import { formatSecurityEngineSend } from './formatSecurityEngine';
import {
  parseTxData,
  parseTxDataTypedData,
  preExecData,
  txData,
  txDataTypedData,
} from './mocks';
import { parseActionSend } from './parseAction';
import {
  parseAction,
  fetchActionRequiredData,
  formatSecurityEngineContext,
} from '../..';
import { parseTypedDataAction } from '../../utils/parseTypedDataAction';

/**
 * https://metamask.github.io/test-dapp/#sendButton
 * Send Eth
 * - [Send Legacy Transaction] button
 */
test.each([
  [parseActionSend, fetchDataSend, formatSecurityEngineSend],
  [parseAction, fetchActionRequiredData, formatSecurityEngineContext],
])('Send -> Transaction', async (_parseAction, _fetchData, _format) => {
  const actionData = _parseAction({
    type: 'transaction',
    data: parseTxData['action'],
    balanceChange: preExecData.balance_change,
    sender: txData.from,
    preExecVersion: preExecData.pre_exec_version,
    gasUsed: preExecData.gas.gas_used,
    tx: txData,
  }) as ParsedTransactionActionData;
  expect(actionData).toMatchSnapshot('parseActionSend');

  const requireData = await _fetchData({
    type: 'transaction',
    actionData,
    contractCall: parseTxData.contract_call,
    chainId: ETH_CHAIN_ID,
    sender: SENDER,
    walletProvider,
    tx: txData,
    apiProvider,
  });

  expect(requireData).toMatchSnapshot('fetchDataSend');

  const ctx = await _format({
    type: 'transaction',
    actionData,
    requireData,
    chainId: ETH_CHAIN_ID,
    isTestnet: false,
    provider: formatProvider,
  });
  expect(ctx).toMatchSnapshot('formatSecurityEngineSend');
});

/**
 * https://syncswap.xyz/
 * gas use USDT
 * ETH -> USDC.e
 */
test.skip.each([
  [
    parseTypedDataAction(parseActionSend),
    fetchDataSend,
    formatSecurityEngineSend,
  ],
  [parseAction, fetchActionRequiredData, formatSecurityEngineContext],
])('Send -> TypedData', async (_parseAction, _fetchData, _format) => {
  const actionData = _parseAction({
    type: 'typed_data',
    data: parseTxDataTypedData['action'],
    typedData: txDataTypedData,
    sender: SENDER,
  });
  expect(actionData).toMatchSnapshot('parseActionSend');

  const requireData = await _fetchData({
    type: 'typed_data',
    actionData,
    chainId: ETH_CHAIN_ID,
    walletProvider,
    apiProvider,
    sender: SENDER,
  });
  expect(requireData).toMatchSnapshot('fetchDataSend');

  const ctx = await _format({
    type: 'typed_data',
    actionData,
    requireData,
    chainId: ETH_CHAIN_ID,
    isTestnet: false,
    provider: formatProvider,
    origin: ORIGIN,
  });
  expect(ctx).toMatchSnapshot('formatSecurityEngineSend');
});
