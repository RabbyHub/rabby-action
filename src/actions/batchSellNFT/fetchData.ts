import { FetchActionRequiredData } from '../../types';
import { fetchDataAssetOrder } from '../assetOrder/fetchData';

export const fetchDataBatchSellNFT: FetchActionRequiredData = async (
  options
) => {
  if (options.type !== 'typed_data') {
    return {};
  }
  if (!options.actionData.contractId || !options.actionData.batchSellNFT) {
    return {};
  }

  return fetchDataAssetOrder(options, {
    receiver: options.actionData.contractId,
  });
};
