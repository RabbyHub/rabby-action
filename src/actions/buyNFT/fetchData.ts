import { FetchActionRequiredData } from '../../types';
import { fetchDataAssetOrder } from '../assetOrder/fetchData';

export const fetchDataBuyNFT: FetchActionRequiredData = async (options) => {
  if (options.type !== 'typed_data') {
    return {};
  }
  if (!options.actionData.contractId || !options.actionData.buyNFT) {
    return {};
  }

  return fetchDataAssetOrder(options, {
    receiver: options.actionData.contractId,
  });
};
