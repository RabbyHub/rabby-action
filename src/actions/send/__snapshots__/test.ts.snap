// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Send: fetchDataSend 1`] = `
{
  "cex": null,
  "contract": null,
  "eoa": {
    "bornAt": 1564655163,
    "id": "0x0c54fccd2e384b4bb6f2e405bf5cbc15a017aafb",
  },
  "hasReceiverMnemonicInWallet": false,
  "hasReceiverPrivateKeyInWallet": true,
  "hasTransfer": true,
  "isTokenContract": false,
  "name": null,
  "onTransferWhitelist": false,
  "protocol": null,
  "receiverIsSpoofing": true,
  "usd_value": 4418.154405584272,
  "usedChains": [],
  "whitelistEnable": false,
}
`;

exports[`Send: fetchDataSend 2`] = `
{
  "cex": null,
  "contract": null,
  "eoa": {
    "bornAt": 1564655163,
    "id": "0x0c54fccd2e384b4bb6f2e405bf5cbc15a017aafb",
  },
  "hasReceiverMnemonicInWallet": false,
  "hasReceiverPrivateKeyInWallet": true,
  "hasTransfer": true,
  "isTokenContract": false,
  "name": null,
  "onTransferWhitelist": false,
  "protocol": null,
  "receiverIsSpoofing": true,
  "usd_value": 4418.154405584272,
  "usedChains": [],
  "whitelistEnable": false,
}
`;

exports[`Send: formatSecurityEngineSend 1`] = `
{
  "send": {
    "cex": null,
    "chainId": "eth",
    "contract": null,
    "hasReceiverMnemonicInWallet": false,
    "hasReceiverPrivateKeyInWallet": true,
    "hasTransfer": true,
    "isTokenContract": false,
    "onTransferWhitelist": false,
    "receiverIsSpoofing": true,
    "to": "0x0c54fccd2e384b4bb6f2e405bf5cbc15a017aafb",
    "usedChainList": [],
  },
}
`;

exports[`Send: formatSecurityEngineSend 2`] = `
{
  "send": {
    "cex": null,
    "chainId": "eth",
    "contract": null,
    "hasReceiverMnemonicInWallet": false,
    "hasReceiverPrivateKeyInWallet": true,
    "hasTransfer": true,
    "isTokenContract": false,
    "onTransferWhitelist": false,
    "receiverIsSpoofing": true,
    "to": "0x0c54fccd2e384b4bb6f2e405bf5cbc15a017aafb",
    "usedChainList": [],
  },
}
`;

exports[`Send: parseActionSend 1`] = `
{
  "send": {
    "to": "0x0c54fccd2e384b4bb6f2e405bf5cbc15a017aafb",
    "token": {
      "amount": 0,
      "chain": "op",
      "credit_score": 100000000,
      "decimals": 18,
      "display_symbol": null,
      "id": "op",
      "is_core": true,
      "is_scam": false,
      "is_suspicious": false,
      "is_verified": true,
      "is_wallet": true,
      "logo_url": "https://static.debank.com/image/coin/logo_url/eth/6443cdccced33e204d90cb723c632917.png",
      "name": "ETH",
      "optimized_symbol": "ETH",
      "price": 2673,
      "price_24h_change": 0.014151838221345444,
      "protocol_id": "",
      "raw_amount": "0",
      "symbol": "ETH",
      "time_at": null,
    },
  },
}
`;

exports[`Send: parseActionSend 2`] = `
{
  "send": {
    "to": "0x0c54fccd2e384b4bb6f2e405bf5cbc15a017aafb",
    "token": {
      "amount": 0,
      "chain": "op",
      "credit_score": 100000000,
      "decimals": 18,
      "display_symbol": null,
      "id": "op",
      "is_core": true,
      "is_scam": false,
      "is_suspicious": false,
      "is_verified": true,
      "is_wallet": true,
      "logo_url": "https://static.debank.com/image/coin/logo_url/eth/6443cdccced33e204d90cb723c632917.png",
      "name": "ETH",
      "optimized_symbol": "ETH",
      "price": 2673,
      "price_24h_change": 0.014151838221345444,
      "protocol_id": "",
      "raw_amount": "0",
      "symbol": "ETH",
      "time_at": null,
    },
  },
}
`;
